rfcs/0054-suspend-mv-on-non-recoverable-errors.md

```diff
@@ -0,0 +1,63 @@
+---
+feature: Suspend MV on Non-Recoverable Errors
+authors:
+  - "Patrick Huang"
+start_date: "2023/2/24"
+---
+# RFC: Suspend MV on Non-Recoverable Errors
+
+# Motivation
+
+Currently we have two error handling mechanisms on MV:
+
+1. Ignore and tolerate the compute error. Relevant issue: https://github.com/risingwavelabs/risingwave/issues/4625)
+2. Report error to meta and trigger a full recovery.
+
+However, when there are non-recoverable errors caused by malform records ingested by the user, the current error handling mechanisms are insufficient:
+
+1. Even if these messages can be skip or fixed via filling in default values, correctness can be affected, especially when sinking to external systems, since the damage may hardly get reverted.
+2. Recovery won’t help here. We may end up with re-consuming the malform records and trigger full recovery over and over again. All MVs in the cluster will be affected and no progress can be made.
+
+This RFC proposes that we should suspend the affected MV when encountering non-recovery errors. Non-recoverable errors by definition cannot be fixed automatically so manual intervention and re-creating the MV are unavoidable. The goals of this RFC are:
+
+- Reduce the radius of impact when non-recoverable errors happen. In other words, non-relevant MVs should not be affected.
+- Improve observability when non-recoverable errors happen. In other words, MV progress and necessary states should be kept to help user/SRE identify the issue.
+
+# Design
+
+## Non-Recoverable Errors
+
+- Errors caused by malform records ingested by the user source. Examples: record too large, record parsing error, record format/type mismatch.
+- Errors caused by failure in sanity check. Examples: state table insert/update/delete sanity check fails.
```

Comment:
> Shall we consider panic as a kind of Non-Recoverable Errors?

+1

Reply:
After some extra thought I think this is bad, because sometimes a `panic` only means the program don't know how to continue but doesn't necessarily means it cann't be recovered from last checkpoint.